### PR TITLE
Refactor rows a little.

### PR DESCRIPTION
**Pull request by chriseth**
The main goal here is to hide the inner structure of a Row so that it can be more easily optimized later.

### CODE REVIEW

File: executor/src/witgen/rows.rs [hunk 1]
```diff
@@ -195,9 +201,63 @@ impl<T: FieldElement> Debug for Cell<'_, T> {
...
+pub fn value_is_known<T: FieldElement>(row: &Row<'_, T>, poly_id: &PolyID) -> bool {
+    row[poly_id].value.is_known()
+}
+
+// TODO will be impl method of row later.
+pub fn set_cell_unknown<T: FieldElement>(row: &mut Row<'_, T>, poly_id: &PolyID) {
+    row[poly_id].value = CellValue::Unknown;
+}
+
+// TODO will be impl method of row later.
```
Comment (georgwiese) [inline executor/src/witgen/rows.rs:241]: What's us from doing that now? Also for the functions below.

File: executor/src/witgen/rows.rs [hunk 2]
```diff
@@ -195,9 +201,63 @@ impl<T: FieldElement> Debug for Cell<'_, T> {
     }
 }
 
-/// A row of cells, indexed by polynomial ID.
 pub type Row<'a, T> = WitnessColumnMap<Cell<'a, T>>;
 
+/// Merges two rows, updating the first.
+/// Range constrainst from the second row are ignored.
```
Comment (georgwiese) [inline executor/src/witgen/rows.rs:207]:
```suggestion
/// Range constraints from the second row are ignored.
```

File: executor/src/witgen/rows.rs [hunk 3]
```diff
@@ -195,9 +201,63 @@ impl<T: FieldElement> Debug for Cell<'_, T> {
...
+        .zip(row2.values())
+        .all(|(cell1, cell2)| cell1.value == cell2.value)
+}
+
+// TODO will be impl method of row later.
+// TODO we need to get rid of column_ids and ensure that the vector is already in that shape.
+pub fn finalize_row<'a, T: FieldElement>(
+    row: &Row<'_, T>,
+    column_ids: impl IntoIterator<Item = &'a PolyID>,
+) -> (Vec<T>, BitVec) {
```
Comment (georgwiese) [inline executor/src/witgen/rows.rs:254]: You could introduce a `FinalizedRow` struct here and use it in `FinalizableData`.

File: executor/src/witgen/rows.rs [hunk 4]
```diff
@@ -195,9 +201,63 @@ impl<T: FieldElement> Debug for Cell<'_, T> {
     }
 }
 
-/// A row of cells, indexed by polynomial ID.
 pub type Row<'a, T> = WitnessColumnMap<Cell<'a, T>>;
 
+/// Merges two rows, updating the first.
+/// Range constrainst from the second row are ignored.
+/// @returns false if there was a conflict (two different known values)
```
[the remark this answers is not in the document]
Comment (chriseth) [reply executor/src/witgen/rows.rs:208]: Right, I can change it back now that I have two iterations.

File: executor/src/witgen/rows.rs (hunk 1 above)
Comment (chriseth) [reply executor/src/witgen/rows.rs:241]: Row is currently only a typedef.

**Review by georgwiese — APPROVED** [review]

**Pull request merged**